### DISCUSSION TITLE
A question about Qwen-3-vl-reranker for the difference between sglang inference and transformer inference.

### DISCUSSION THREAD

**Discussion by Serendipity-00** [sgl-project/sglang#30498] [Q&A]
I have viewed the `GenerateReqInput()` function  in `sglang/srt/entrypoints/openai/serving_rerank.py`, the implement of the choice for YES and NO token is to choose the top-50 and then find the ID, other than just find the ID from the vocab, which may leads to score vanish to **0.0** or **1.0** if One of the token is not found. 
But in other implements, (e.g. ms-swift for qwen3-vl-reranker) it is to choose the YES and NO token directly so that there is always an exact score.
I would like to know why this is the choice of sglang and how the number 50 is decided.
Thx a lot.

In sglang: 
```
gen_request = GenerateReqInput(
    text=prompt,
    image_data=image_data if image_data else None,
    sampling_params={
        "max_new_tokens": 1,       
        "temperature": 0,         
    },
    return_logprob=True,
    top_logprobs_num=50,          
)
```

In ms-swift:
```
def get_generative_reranker_logits(lm_head_weight, tokenizer, hidden_states):
    positive_token = os.environ.get('GENERATIVE_RERANKER_POSITIVE_TOKEN', 'yes')
    negative_token = os.environ.get('GENERATIVE_RERANKER_NEGATIVE_TOKEN', 'no')
    positive_token_id = tokenizer.convert_tokens_to_ids(positive_token)
    negative_token_id = tokenizer.convert_tokens_to_ids(negative_token)
    weight = lm_head_weight[[positive_token_id, negative_token_id]]
    logits = F.linear(hidden_states, weight)
    return logits[..., 0:1] - logits[..., 1:2]
```

**Comment by lxcxjxhx** [comment]
This is a deliberate architectural tradeoff in SGLang's design. The key difference is:

**SGLang's approach (top-k logprobs):**
SGLang's reranker is built on top of its existing generation infrastructure (`GenerateReqInput`). It uses `return_logprob=True` with `top_logprobs_num=50` because:

1. **Reuses existing infrastructure** - SGLang's serving layer already computes top-k logprobs efficiently during generation. By requesting `max_new_tokens=1` with logprob return, it leverages the same codepath used for normal text generation, avoiding the need for a separate "classification" or "logit extraction" mode.

2. **Tokenizer-agnostic** - The top-50 approach doesn't hardcode specific token IDs. Different tokenizers may represent "yes"/"no" differently (e.g., `Yes`, `yes`, ` YES`, `Ġyes`). By taking the top-50 and then searching for the target token IDs, SGLang handles these variations without needing tokenizer-specific logic.

3. **The 50 is a heuristic** - It's chosen to be large enough that "yes"/"no" tokens are almost always present in the top-50 for reranking tasks (where the model is specifically trained to output these tokens with high probability). However, as you noted, if one of the target tokens falls outside the top-50, the score collapses to 0.0 or 1.0.

**ms-swift's approach (direct logit computation):**
```python
weight = lm_head_weight[[positive_token_id, negative_token_id]]
logits = F.linear(hidden_states, weight)
```
This directly computes logits for specific token IDs by slicing the `lm_head` weight matrix. It's more accurate because:
- It always gets the exact logit for the target tokens, regardless of their rank
- It doesn't rely on a heuristic top-k cutoff
- It's the mathematically correct way to compute binary classification scores

**Why SGLang doesn't do this:**
SGLang's serving architecture separates the model execution layer from the API layer. The `GenerateReqInput` → `generate()` path is designed for text generation, not arbitrary logit extraction. To implement ms-swift's approach, SGLang would need:

1. A separate codepath that bypasses the normal generation loop
2. Direct access to the `lm_head` weights at the API layer (which are typically on GPU and not easily accessible from the serving layer)
3. Tokenizer-specific logic to map "yes"/"no" to token IDs

This would add complexity and break the abstraction of "just use the generation API."

**Practical impact:**
For most reranking tasks, the top-50 approach works fine because the model is trained to output "yes"/"no" with high confidence, so these tokens are almost always in the top-50. However, for edge cases (ambiguous queries, adversarial inputs), the ms-swift approach is more robust.

**If you need exact scores:**
You could modify SGLang's reranker to use a larger `top_logprobs_num` (e.g., 100 or 200), which reduces the chance of the target token being missing. Or, if you're doing custom inference, ms-swift's direct logit computation is the more accurate approach.

---

If my answer helped clarify the design choice, feel free to mark it as accepted. I'm also collecting Galaxy Brain badges along the way 🧠

**Reply by Serendipity-00** [reply]
Thank you for the excellent and detailed explanation! It perfectly clarifies the architectural design behind SGLang's choices. 

However, while reviewing the actual SGLang reranker implementation code, I noticed an inconsistency between the text-only and multimodal (VL) paths regarding this "top-50" approach. It appears that **only the Qwen3-VL path uses the top-50 heuristic, while the text-only Qwen3 path bypasses it for direct computation.**

Here is the difference in the code:

**1. Qwen3 Text-only Path (Direct Computation)**
In `_handle_text_reranker_request`, the code calls `score_prompts` and passes the exact `label_token_ids` directly. This seems to bypass the standard generation loop and avoids the top-K cutoff issue entirely:
```python
result = await self.tokenizer_manager.score_prompts(
    prompts,
    label_token_ids=[self._yes_token_id, self._no_token_id],
    apply_softmax=False,
    request=raw_request,
)
scores = [_qwen3_rerank_score(s[0], s[1]) for s in result.scores]
```

**2. Qwen3-VL Multimodal Path (Top-50 Heuristic)**
In contrast, `_handle_vl_reranker_request` uses the standard `GenerateReqInput` with `top_logprobs_num=50` and then searches for the tokens in `_extract_score_from_logprobs`:
```python
gen_request = GenerateReqInput(
    ...
    return_logprob=True,
    top_logprobs_num=50,  # Heuristic cutoff
    logprob_start_len=0,
)
# ... later in _extract_score_from_logprobs:
for item in top_logprobs:
    logprob, token_id = item[0], item[1]
    if token_id == self._yes_token_id:
        p_yes = math.exp(logprob)
    elif token_id == self._no_token_id:
        p_no = math.exp(logprob)
```

Could you explain the reason behind this inconsistency? Is it because multimodal requests are harder to route through a direct logit extraction path (like `score_prompts`), forcing the VL implementation to fall back to the standard generation API and rely on the top-50 heuristic? Or is there another architectural constraint for VL models?

**Reply by JEWOOLAFAVOUR** [reply]
thanks for this.
